openwebstart/src/main/java/com/openwebstart/app/icon/ApplicationIconCache.java

```diff
@@ -0,0 +1,128 @@
+package com.openwebstart.app.icon;
+
+import com.openwebstart.app.Application;
+import net.adoptopenjdk.icedteaweb.Assert;
+
+import java.awt.image.BufferedImage;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class ApplicationIconCache {
+
+    private final Map<IconDescription, BufferedImage> cache = new HashMap<>();
+
+    private final Map<IconDescription, CompletableFuture<BufferedImage>> inProgressMap = new HashMap<>();
+
+    private final Lock inProgressMapLock = new ReentrantLock();
+
+    private final Lock cacheLock = new ReentrantLock();
+
+    private final Executor executor;
+
+    public ApplicationIconCache(final Executor executor) {
+        this.executor = Assert.requireNonNull(executor, "executor");
+    }
+
+    public synchronized Optional<BufferedImage> get(final Application application, final IconDimensions dimension) {
+        final IconDescription description = new IconDescription(application.getId(), dimension);
+        cacheLock.lock();
+        try {
+            return Optional.ofNullable(cache.get(description));
+        } finally {
+            cacheLock.unlock();
+        }
+    }
+
+    public synchronized CompletableFuture<BufferedImage> getOrLoadIcon(final Application application, final IconDimensions dimension) {
+        final Optional<BufferedImage> fromCache = get(application, dimension);
+        if (fromCache.isPresent()) {
+            return CompletableFuture.completedFuture(fromCache.get());
+        } else {
+            return loadAndAdd(application, dimension);
+        }
+    }
+
+    public synchronized void triggerDownload(final Application application, final IconDimensions dimension, final boolean reloadIfAlreadyInCache) {
+        if (reloadIfAlreadyInCache || !isInCache(application, dimension)) {
+            executor.execute(() -> loadAndAdd(application, dimension));
```

Comment:
Is it necessary to run `loadAndAdd` in a separate thread as the method will start yet another thread...

Reply:
yes it is:

triggerDownload is called by the UI. Therefore I should never block. While loadAndAdd internally starts an extra thread for the download it creates a lock directly at the start. Based on this the loadAndAdd method can block and should not be called directly from the EDT